sdk/dotnet/Pulumi.Automation/PulumiFn.ServiceProvider.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Copyright 2016-2021, Pulumi Corporation
+
+using System;
+using System.Reflection;
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Pulumi.Automation
+{
+    internal class PulumiFnServiceProvider : PulumiFn
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly Type _stackType;
+
+        public PulumiFnServiceProvider(
+            IServiceProvider serviceProvider,
+            Type stackType)
+        {
+            if (serviceProvider is null)
+                throw new ArgumentNullException(nameof(serviceProvider));
+
+            if (stackType is null)
+                throw new ArgumentNullException(nameof(stackType));
+
+            var pulumiStackType = typeof(Pulumi.Stack);
+            if (!pulumiStackType.IsAssignableFrom(stackType) || pulumiStackType == stackType)
```

Comment:
So what happens if `pulumiStackType == stackType`, that is, `new PulumiFnServiceProvider(serviceProvider, typeof(Pulumi.Stack))` - we currently would throw here. Is that as intended? Sorry, I have not used inline programs before so building up my intuition here.

Reply:
I decided to throw for that because in practice if you provide `Pulumi.Stack` than your `up` operation would be a no-op and wouldn't declare any resources, so it seemed like the right thing to do. I figured we'd want to enforce that the provided type be a derivation of `Pulumi.Stack` and not just assignable to `Pulumi.Stack`, but we could change that

Reply:
but I guess if we wanted parity with the generic API we wouldn't throw for using `Pulumi.Stack`, since you could do `PulumiFn.Create<Pulumi.Stack>()` without throwing.. so we can revert that if you want

Reply:
That's reasonable, perfect so this line is as intended.